### ISSUE TITLE
[FEATURE] Handling Persistent Requests and Thread Context Issues in SSE Interfaces Using langchain4j

### ISSUE THREAD

**Issue by blackhumors** [langchain4j/langchain4j#1551]
1. When I developed an SSE interface using langchain4j, the StreamingChatLanguageModel.generate continues to run even after the webpage disconnects the SSE request. How can I immediately stop the requests to the OpenAI server?
2. The server's logs are chain-based and rely on a correlation ID stored in ThreadLocal for quick retrieval of logs related to the current request. However, there is currently no method to pass the ThreadLocal content to the thread running StreamingChatLanguageModel.generate. Do you have any suggestions on how to handle this?

**Comment by InAnYan** [comment]
Hi, @wangyulong-patsnap! Regarding your first question, take a look at my issue: https://github.com/langchain4j/langchain4j/issues/1454, maybe it could give you some ideas to solve your problem

**Comment by blackhumors** [comment]
> Hi, @wangyulong-patsnap! Regarding your first question, take a look at my issue: #1454, maybe it could give you some ideas to solve your problem

I don't think this is a good solution because you can't tell when the page's SSE disconnects. I would prefer to have an option that allows the user to control it.

**Comment by langchain4j** [comment]
HI @wangyulong-patsnap 

> When I developed an SSE interface using langchain4j, the StreamingChatLanguageModel.generate continues to run even after the webpage disconnects the SSE request. How can I immediately stop the requests to the OpenAI server?

This is a known "problem" of okhttp (which we use under the hood). okhttp client maintains a thread pool and threads are cached for 1 minute. Since we are slowly migrating from okhttp, this issue should be resolved soon.

> The server's logs are chain-based and rely on a correlation ID stored in ThreadLocal for quick retrieval of logs related to the current request. However, there is currently no method to pass the ThreadLocal content to the thread running StreamingChatLanguageModel.generate. Do you have any suggestions on how to handle this?

Do I understand correctly that you want to access `ThreadLocal` from withing the `onNext`, `onComplete`, etc methods of `StreamingResponseHandler`?

**Comment by blackhumors** [comment]
> Do I understand correctly that you want to access ThreadLocal from withing the onNext, onComplete, etc methods of StreamingResponseHandler?

Yes, I need to pass the business layer's ThreadLocal into methods such as onNext and onComplete to facilitate logging with the entire request chain identifier, making it easier to troubleshoot business issues

**Comment by langchain4j** [comment]
You can store the value of a `ThreadLocal` before executing `StreamingChatLanguageModel.generate()`:

```java
            ThreadLocal<String> threadLocal = new ThreadLocal<>();
            threadLocal.set("test");

            // ...
            
            String threadLocalValue = threadLocal.get();

            model.generate(messages, new StreamingResponseHandler<AiMessage>() {

                @Override
                public void onNext(String token) {
                    System.out.println("Thread local: " + threadLocalValue);
                }

                @Override
                public void onComplete(Response<AiMessage> response) {
                    System.out.println("Thread local: " + threadLocalValue);
                }

                @Override
                public void onError(Throwable error) {
                    System.out.println("Thread local: " + threadLocalValue);
                }
            });
```

**Comment by blackhumors** [comment]
> You can store the value of a `ThreadLocal` before executing `StreamingChatLanguageModel.generate()`:
> 
> ```java
>             ThreadLocal<String> threadLocal = new ThreadLocal<>();
>             threadLocal.set("test");
> 
>             // ...
>             
>             String threadLocalValue = threadLocal.get();
> 
>             model.generate(messages, new StreamingResponseHandler<AiMessage>() {
> 
>                 @Override
>                 public void onNext(String token) {
>                     System.out.println("Thread local: " + threadLocalValue);
>                 }
> 
>                 @Override
>                 public void onComplete(Response<AiMessage> response) {
>                     System.out.println("Thread local: " + threadLocalValue);
>                 }
> 
>                 @Override
>                 public void onError(Throwable error) {
>                     System.out.println("Thread local: " + threadLocalValue);
>                 }
>             });
> ```

This is how I currently implement it: I have customized a `StreamingResponseHandler`. Initializing this `StreamingResponseHandler` requires transferring all the contents from `ThreadLocal` one by one into the `StreamingResponseHandler`. It feels somewhat inelegant, but there seems to be no better method.

**Comment by sakoojiang** [comment]
I developed an SSE use this way:

# chat code
```java
    public Flux<AiResult<AiMessage>> stream(ChatRequest chatRequest) {
        IntentResult intentResult=this.intentRecognition(chatRequest);
        List<ToolExecution> toolExecutions= Lists.newArrayList();
        List<Content> documents=Lists.newArrayList();
        StreamingAssistant assistant=this.aiServices(chatRequest, StreamingAssistant.class);
        TokenStream tokenStream=assistant.generate(chatRequest.getMessage());
        Sinks.Many<AiResult<AiMessage>> sink= Sinks.many().unicast().onBackpressureBuffer();
        tokenStream.onNext(token -> {
                    AiMessage message=AiMessage.from(token);
                    AiResult<AiMessage> ret=new AiResult<>(message);
                    sink.tryEmitNext(ret);
                }).onComplete(response -> {
                    FinishReason finishReason=response.finishReason()!=null?response.finishReason(): FinishReason.STOP;
                    AiResult<AiMessage> ret=new AiResult<>(AiMessage.from(""),response.tokenUsage(),documents,finishReason,toolExecutions,intentResult);
                    sink.tryEmitNext(ret);
                    sink.tryEmitComplete();
                }).onRetrieved(documents::addAll)
                .onToolExecutor(toolExecutions::add)
                .onError(throwable -> {
                    log.error("streaming error",throwable);
                    sink.tryEmitError(throwable);
                }).start();
        return  sink.asFlux();


    }
```

# controller code
```java
  @PostMapping("/ai/generate")
    public Flux<ServerSentEvent<ChatResponseModel>> stream(@RequestBody ChatRequest chatRequest){
        return generate(chatRequest)
                .flatMapIterable(ChatResponseModel::of)
                .onErrorResume(error -> {
                    log.error("",error);
                    return  Flux.just(ChatResponseModel.error(error.getMessage()));

                })
                .map(responseModel -> ServerSentEvent.builder(responseModel)
                        .build())

                ;
    }
     private Flux<AiResult<AiMessage>> generate(ChatRequest chatRequest){
        try{
            if(chatRequest.isStream()){
                return  stream(chatRequest);
            }else{
                return Flux.just(call(chatRequest));
            }
        }catch (Throwable e){
            log.error("",e);
            return Flux.error(e);
        }


    }


```

**Comment by langchain4j** [comment]
@sakoojiang perharps you could simplify it by using `Flux` [support of LC4j](https://docs.langchain4j.dev/tutorials/ai-services#flux), here is an [example](https://github.com/marcushellberg/java-ai-playground/blob/81383f8b6488fe0b9597bc9b849f1962bc728aeb/src/main/java/org/vaadin/marcus/client/AssistantService.java#L19).

**Comment by catofdestruction** [comment]
> @sakoojiang perharps you could simplify it by using `Flux` [support of LC4j](https://docs.langchain4j.dev/tutorials/ai-services#flux), here is an [example](https://github.com/marcushellberg/java-ai-playground/blob/81383f8b6488fe0b9597bc9b849f1962bc728aeb/src/main/java/org/vaadin/marcus/client/AssistantService.java#L19).perharps 你可以通过使用[LC4j 的](https://docs.langchain4j.dev/tutorials/ai-services#flux)`Flux`支持来简化它，这里是一个[例子](https://github.com/marcushellberg/java-ai-playground/blob/81383f8b6488fe0b9597bc9b849f1962bc728aeb/src/main/java/org/vaadin/marcus/client/AssistantService.java#L19)。


Using this simplified example, can the return type only be Flux\<String\>? Is it possible to return Flux\<CustomObject\> to return models or business information, or more content? Thanks

**Comment by langchain4j** [comment]
@catofdestruction no, only `Flux<String>` is supported at the moment, but in the future we will consider supporting `Flux<Pojo>`.
What is your use case for `Flux<Pojo>`?

**Comment by catofdestruction** [comment]
> @catofdestruction no, only `Flux<String>` is supported at the moment, but in the future we will consider supporting `Flux<Pojo>`. What is your use case for `Flux<Pojo>`?

Thank you for your response. I am currently trying to use our @AIService to construct an agent that returns streaming content. In most cases, a single agent returning a Flux\<String\> is sufficient. However, in some scenarios, multiple agents need to collaborate with each other. I hope the sub-agent can not only return streaming content but also provide some custom process information (e.g., tool call results and processes).

Additionally, I would like to inquire if there are any plans to extend the framework with capabilities related to ReAct Agent and Multi-Agent collaboration. Thank you very much.

**Comment by langchain4j** [comment]
> I hope the sub-agent can not only return streaming content but also provide some custom process information (e.g., tool call results and processes).

Please check https://docs.langchain4j.dev/tutorials/ai-services#streaming, there is `onToolExecuted` and `onRetrieved` callbacks.

> Additionally, I would like to inquire if there are any plans to extend the framework with capabilities related to ReAct Agent and Multi-Agent collaboration. Thank you very much.

ReAct: what exactly are you missing? ReAct is pretty much built-in in most LLMs supporting tools nowadays.
Multi-agent: not in the coming 3 months for sure.

**Comment by catofdestruction** [comment]
> > I hope the sub-agent can not only return streaming content but also provide some custom process information (e.g., tool call results and processes).
> 
> Please check https://docs.langchain4j.dev/tutorials/ai-services#streaming, there is `onToolExecuted` and `onRetrieved` callbacks.
> 
> > Additionally, I would like to inquire if there are any plans to extend the framework with capabilities related to ReAct Agent and Multi-Agent collaboration. Thank you very much.
> 
> ReAct: what exactly are you missing? ReAct is pretty much built-in in most LLMs supporting tools nowadays. Multi-agent: not in the coming 3 months for sure.

Thank you very much for your response, it helped me a lot.

**Comment by sakoojiang** [comment]
> 
> Please check https://docs.langchain4j.dev/tutorials/ai-services#streaming, there is `onToolExecuted` and `onRetrieved` callbacks.
> 



OnToolExecuted doesn't seem to exist in version 0.35.0, should it be integrated in the next version?
I remember I submit the pull request for onToolExecuted after the release of version 0.35.0.

[https://github.com/langchain4j/langchain4j/pull/1818#issue-2541759268](url)

But I feel a bit surprised to see that there is already content about onToolExecuted in the current document

**Comment by langchain4j** [comment]
@sakoojiang indeed, documentation was published prematurely. 0.36.0 will be released this week, probably on Wednesday
